### ISSUE TITLE
Bug in no-extra-bind

### ISSUE THREAD

**Issue by mjomble** [eslint/eslint#5058]
I found another bug in this rule.

If the `this` keyword only appears as the parameter to `bind`, it doesn't get registered in the right scope because we've already exited the `FunctionExpression`, but not yet from its associated `CallExpression`.

So in code like this:

```javascript
var a = function() {
  (function() {
    this.b
  }.bind(this))
}.bind(c)
```

the outer `bind(c)` is marked as unnecessary, even though `this` is actually used in the bound function.

The bug gets fixed (and other tests will also pass) if I replace this:

```javascript
if (top.depth === 0) {
    top.found++;
}
```

with this:

```javascript
if (top.depth === 0) {
    top.found++;
} else if (top.depth === -1 && scope.length > 1) {
    var parentOfTop = scope[scope.length - 2];
    parentOfTop.found++;
}
```

It feels slightly hacky, but I'm not sure how to solve it in a cleaner way.

I can submit a PR with this fix if needed.

**Comment by eslintbot** [comment]
@mjomble Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by mysticatea** [comment]
Thank you for this issue!

I confirmed it.

Hmm, I'm hoping to refactor this rule...
I think scope's information should be added to the stack at function nodes.

Also, I found 2 bugs in addition in this rule right now.
The following patterns should be not warned, but those are warned:

```js
(function() {}[bind](abc)); // computed property.
(function() { (() => foo(this)); }.bind(abc)); // `this` in an arrow is upper scope's
```

May I do it?

**Comment by mjomble** [comment]
Sounds good to me :)

**Comment by mysticatea** [comment]
Thank you.
And thank you for investigating!
